## js 中的异步种类

- 第一种最常见的是使用回调函数的方式，使用回调函数的方式有一个缺点是，多个回调函数嵌套的时候会造成回调函数地狱，上下两层的回调函数间的代码耦合度太高，不利于代码的可维护。

- 第二种使用事件监听，比如 JQuery 的 trigger 和 on，trigger 暴露事件 on 监听事件。每次使用还得注册事件监听再进行触发挺麻烦的，代码也不太优雅。

- 第三种是 Promise 的方式，使用 Promise 的方式可以将嵌套的回调函数作为链式调用。但是使用这种方法，有时会造成多个 then 的链式调用，可能会造成代码的语义不够明确。Promise 对象一旦新建就会立即执行，无法中途取消。

- 第四种是使用 generator 的方式，它可以在函数的执行过程中，将函数的执行权转移出去，在函数外部我们还可以将执行权转移回来。当我们遇到异步函数执行的时候，将函数执行权转移出去，当异步函数执行完毕的时候我们再将执行权给转移回来。因此我们在 generator 内部对于异步操作的方式，可以以同步的顺序来书写。使用这种方式我们需要考虑的问题是何时将函数的控制权转移回来，因此我们需要有一个自动执行 generator 的机制，比如说 co 模块等方式来实现 generator 的自动执行。

* 第五种是使用 async 函数的形式，async 函数是 generator 和 promise 实现的一个自动执行的语法糖，它内部自带执行器，当函数内部执行到一个 await 语句的时候，如果语句返回一个 promise 对象，那么函数将会等待 promise 对象的状态变为 resolve 后再继续向下执行。因此我们可以将异步逻辑，转化为同步的顺序来书写，并且这个函数可以自动执行。

## promise

### promise 特点

- promise 解决层层回调函数，代码不清晰问题，使异步代码能链式调用 流程更加清晰，代码更加优雅
- 一个 Promise 对象值是未知的，状态是可变的，但是无论怎么变化，它的状态永远处于以下三种之间：
  - pending：初始状态，既不是成功，也不是失败。
  - fulfilled：意味着操作成功完成。
  - rejected：意味着操作失败。
- Promise 的状态会发生变化，成功时会从 pending -> fulfilled，失败时会从 pending -> rejected，但是此过程是不可逆的，也就是不能从另外两个状态变成 pending。fulfilled/rejected 这两个状态也被称为 settled 状态。
- Promise 的构造函数是同步执行的。then 是异步执行的。
- 无法取消 Promise

### promise 方法

- then(onFulfilled[, onRejected])
  - onFulfilled 当 Promise 变成已完成状态(fulfilled)时调用的回调函数
  - onRejected 当 Promise 变成接受状态或拒绝状态(rejected)时调用的回调函数。当前的 err 不会进入，会进入下一个 then 方法里面的 onRejected 方法里面。不常用，我们一般使用 catch。
- catch() 获取的是出错的异常
- finally() 不管成功失败都会执行

静态方法

- resolve() 表示成功 返回 promise 对象 在 reject 里面用到 resolve 方法 可以使用 Promise.resolve('success')
- reject() 表示失败 返回 promise 对象 在 resolve 里面用到 reject 方法 可以使用 Promise.reject('error')
- all([p1, p2]) 接收一个 Promise 对象数组作为参数,参数所有回调成功才是成功，返回值数组与参数顺序一致,参数数组其中一个失败，则触发失败状态，进入 catch。第一个触发失败的 Promise 错误信息作为 Promise.all 的错误信息。
- race([p1, p2]) 接收一个 Promise 对象数组作为参数,参数里的任意一个子 promise 被成功或失败后，则触发失败状态，进入 catch。父 promise 马上也会用子 promise 的成功返回值或失败详情作为结果。
- allSettled([p1, p2]) 接收一个 Promise 对象数组作为参数,不管子 promise 成功失败都返回。

## Generator

Generators 是可以用来控制迭代器的函数。它们可以暂停，然后在任何时候恢复。

这个是 Generator 的定义方法，有几个点值得注意：

1. 比普通函数多一个 `*`
2. 函数内部用 yield 来控制程序的执行的“暂停”
3. 函数的返回值通过调用 next 来“恢复”程序执行

Generator 函数的定义不能使用箭头函数，否则会触发 SyntaxError 错误

```js
function* generatorFn() {
  console.log(1);
  yield console.log(2);
  console.log(3);
  yield console.log(4);
  console.log(5);
}

let ge = generatorFn();
ge.next(); // 1 2
ge.next(); // 3 4
ge.next(); // 5

// yield命令后面如果不加星号，返回的是整个数组，加了星号就表示返回的是数组的遍历器
function* gen1() {
  yield ["a", "b", "c"];
}
for (let val of gen1()) {
  console.log(val);
}
// ["a", "b", "c"]

// ------------------- 上下分割

function* gen2() {
  yield* ["a", "b", "c"];
}
for (let val of gen2()) {
  console.log(val);
}
// a b c
```

### generator 传参

```js
function* gen1(x) {
  let y = 2 * (yield x + 1);
  let z = yield y / 3;
  return x + y + z;
}

const g = gen1(5);
console.log(g.next()); // 6
console.log(g.next()); // NaN
console.log(g.next()); // NaN

function* gen2(x) {
  let y = 2 * (yield x + 1); // 2 * 12 = 24
  let z = yield y / 3; // z=22
  return x + y + z; // 5 + 24 + 22
}

const g2 = gen2(5);
console.log(g2.next()); // 6
console.log(g2.next(12)); // 8 参数作为上一个yield的返回值
console.log(g2.next(22)); // 51
```

### 模拟请求

```js
function request(url) {
  ajax(url, (res) => {
    getData.next(res);
  });
}

function* gen() {
  let res1 = yield request("static/a.json");
  console.log(res1); // a.json文件内容
  let res2 = yield request("static/b.json");
  console.log(res2); // b.json文件内容
  let res3 = yield request("static/c.json");
  console.log(res3); // c.json文件内容
}
let getData = gen();
getData.next();
```

### 取 7 的倍数

```js
function* count(x = 1) {
  while (true) {
    if (x % 7 === 0) {
      yield x;
    }
    x++;
  }
}
// es5中就是个死循环 因为es5的循环需要有个终止值，但我们这个需求没有终止，一直在数数
let n = count();
console.log(n.next().value); // 7
console.log(n.next().value); // 14
console.log(n.next().value); // 21
console.log(n.next().value); // 28
console.log(n.next().value); // 35
console.log(n.next().value); // 42
```

## async await

没有了链式调用，用完全同步的方法写异步代码 缺点是没有 promise.all()这种并发的方法

## 常见问题

### 什么是 Promise 对象，什么是 Promises/A+ 规范？

Promise 对象是异步编程的一种解决方案，最早由社区提出。Promises/A+ 规范是 JavaScript Promise 的标准，规定了一个 Promise 所必须具有的特性。

Promise 是一个构造函数，接收一个函数作为参数，返回一个 Promise 实例。一个 Promise 实例有三种状态，分别是 pending、resolved 和 rejected，分别代表了进行中、已成功和已失败。实例的状态只能由 pending 转变 resolved 或者 rejected 状态，并且状态一经改变，就凝固了，无法再被改变了。状态的改变是通过 resolve() 和 reject() 函数来实现的，我们可以在异步操作结束后调用这两个函数改变 Promise 实例的状态，它的原型上定义了一个 then 方法，使用这个 then 方法可以为两个状态的改变注册回调函数。这个回调函数属于微任务，会在本轮事件循环的末尾执行。

```js
// 手写一个 Promise
const PENDING = "pending";
const RESOLVED = "resolved";
const REJECTED = "rejected";

function MyPromise(fn) {
  // 保存初始化状态
  var self = this;

  // 初始化状态
  this.state = PENDING;

  // 用于保存 resolve 或者 rejected 传入的值
  this.value = null;

  // 用于保存 resolve 的回调函数
  this.resolvedCallbacks = [];

  // 用于保存 reject 的回调函数
  this.rejectedCallbacks = [];

  // 状态转变为 resolved 方法
  function resolve(value) {
    // 判断传入元素是否为 Promise 值，如果是，则状态改变必须等待前一个状态改变后再进行改变
    if (value instanceof MyPromise) {
      return value.then(resolve, reject);
    }

    // 保证代码的执行顺序为本轮事件循环的末尾
    setTimeout(() => {
      // 只有状态为 pending 时才能转变，
      if (self.state === PENDING) {
        // 修改状态
        self.state = RESOLVED;

        // 设置传入的值
        self.value = value;

        // 执行回调函数
        self.resolvedCallbacks.forEach((callback) => {
          callback(value);
        });
      }
    }, 0);
  }

  // 状态转变为 rejected 方法
  function reject(value) {
    // 保证代码的执行顺序为本轮事件循环的末尾
    setTimeout(() => {
      // 只有状态为 pending 时才能转变
      if (self.state === PENDING) {
        // 修改状态
        self.state = REJECTED;

        // 设置传入的值
        self.value = value;

        // 执行回调函数
        self.rejectedCallbacks.forEach((callback) => {
          callback(value);
        });
      }
    }, 0);
  }

  // 将两个方法传入函数执行
  try {
    fn(resolve, reject);
  } catch (e) {
    // 遇到错误时，捕获错误，执行 reject 函数
    reject(e);
  }
}

MyPromise.prototype.then = function(onResolved, onRejected) {
  // 首先判断两个参数是否为函数类型，因为这两个参数是可选参数
  onResolved =
    typeof onResolved === "function"
      ? onResolved
      : function(value) {
          return value;
        };

  onRejected =
    typeof onRejected === "function"
      ? onRejected
      : function(error) {
          throw error;
        };

  // 如果是等待状态，则将函数加入对应列表中
  if (this.state === PENDING) {
    this.resolvedCallbacks.push(onResolved);
    this.rejectedCallbacks.push(onRejected);
  }

  // 如果状态已经凝固，则直接执行对应状态的函数

  if (this.state === RESOLVED) {
    onResolved(this.value);
  }

  if (this.state === REJECTED) {
    onRejected(this.value);
  }
};
```
